test/py/test_quiesce.py

```diff
@@ -195,6 +195,31 @@ def test_allowed_funcs_in_quiesced_dma_unregister_busy(mock_quiesce,
     mock_dma_unregister.assert_called_once_with(ctx, mock.ANY)
 
 
+@patch('libvfio_user.migr_trans_cb', side_effect=_side_effect)
+@patch('libvfio_user.quiesce_cb')
+def test_allowed_funcs_in_quiesed_migration(mock_quiesce,
+                                            mock_trans):
+
+    global ctx, sock
+    _map_dma_region(ctx, sock)
+    transition_to_state(ctx, sock, VFIO_USER_DEVICE_STATE_STOP)
```

Comment:
This implies that STOP is the only state we can transition to once quiesced. If this is true then we should have negative tests for all other migration states, no?

Reply:
I have to be honest I don't really understand how this works - I just copied across the test from the previous version and changed it to the migration v2 API. I also changed the previous test's transition to STOP_COPY to a transition to STOP because of the following assertion that transition is called once, and it gets called multiple times because of the previous state if we transition to STOP_COPY.

In the actual code (both in v1 and v2) we allow any state transition to happen once quiesced.

Reply:
> I just copied across the test from the previous version

Do you mean migration v1?

Reply:
Yes

Reply:
I `git diff`ed against master and can't find such a test?

Reply:
https://github.com/nutanix/libvfio-user/blob/master/test/py/test_quiesce.py#L200

Reply:
Right, why does it show that you're adding this function though?

Reply:
nvm, github being github, need to manually configure the base commit to show the right thing

Reply:
> and it gets called multiple times because of the previous state if we transition to STOP_COPY.

Not sure I follow, is it executed mulitple times because of the intermediate transitions?

I realize that the base code isn't well-written. Ideally we should refactor it into a single function that attempts to execute allowed/disallowed API calls during queisce and see whether it was permitted or not.

Reply:
Surreal discussion, no actual issue here.